docs/architecture/adr-050-sign-mode-textual-annex1.md

```diff
@@ -0,0 +1,218 @@
+  - alternatively, we can decide to add a protobuf option to denote messages that are `sdk.Msg`s.
+- Field names follow sentence case
+  - replace `_` with a spaces
+  - capitalize first letter
+- Field names are ordered by their Protobuf field number
+- Nesting:
+  - if a field contains a nested message, we value-render the underlying message, and render the template:
+  ```
+  <field_name>: <1st line of value-rendered message>
+  > <lines 2-n of value-rendered message>             // Notice the `>` prefix.
+  ```
+  - `>` character used to represent nesting. For each additional level of nesting, add `>`.
+
+#### Examples
+
+Given the following Protobuf messages:
+
+```proto
+enum VoteOption {
+  VOTE_OPTION_UNSPECIFIED = 0;
+  VOTE_OPTION_YES = 1;
+  VOTE_OPTION_ABSTAIN = 2;
+  VOTE_OPTION_NO = 3;
+  VOTE_OPTION_NO_WITH_VETO = 4;
+}
+
+message WeightedVoteOption {
+  VoteOption option = 1;
+  string     weight = 2 [(cosmos_proto.scalar) = "cosmos.Dec"];
+}
+
+message Vote {
+  uint64 proposal_id = 1;
+  string voter       = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  reserved 3;
+  repeated WeightedVoteOption options = 4;
+}
+```
+
+we get the following encoding for the `Vote` message:
+
+```
+Vote object
+> Proposal id: 4
+> Vote: cosmos1abc...def
+> Options: 2 WeightedVoteOptions
+> Options (1/2): WeightedVoteOption object
+>> Option: Yes
+>> Weight: 0.7
+> Options (2/2): WeightedVoteOption object
+>> Option: No
+>> Weight: 0.3
+> End of Options
+```
+
+### Enums
+
+- String case convention: snake case to sentence case
+- Allow optional annotation for textual name (TBD)
+- Algorithm:
+  - convert enum name (`VoteOption`) to snake_case (`VOTE_OPTION`)
+  - truncate that prefix + `_` from the enum name if it exists (`VOTE_OPTION_` gets stripped from `VOTE_OPTION_YES` -> `YES`)
+  - convert rest to sentence case: `YES` -> `Yes`
+  - in summary: `VOTE_OPTION_YES` -> `Yes`
+
+#### Examples
+
+See example above with `message Vote{}`.
+
+### `google.protobuf.Any`
+
+- Applies to `google.protobuf.Any`
+- Rendered as:
+
+```
+<type_url>
+> <value rendered underlying message>
+```
+
+#### Examples
+
+TODO
+
+### `google.protobuf.Timestamp` (TODO)
+
+Rendered as either ISO8601 (`2021-01-01T12:00:00Z`) or a more standard English-language date format (`Jan. 1, 2021 12:00 UTC`)
+
+### `google.protobuf.Duration` (TODO)
+
+- rendered in terms of weeks, days, hours, minutes and seconds as these time units can be measured independently of any calendar and duration values are in seconds (so months and years can't be used precisely)
+- total seconds values included at the end so users have both pieces of information
+- Ex:
+  - `1483530 seconds` -> `2 weeks, 3 days, 4 hours, 5 minutes, 30 seconds (1483530 seconds total)`
+
+### address bytes
+
+We currently use `string` types in protobuf for addresses so this may not be needed, but if any address bytes are used in sign mode textual they should be rendered with bech32 formatting
+
```

Comment:
I would add more type cases:

* null values
* strings: must be encoded using escape characters, maybe we should wrap them in `"..."` to distinguish from bytes and numbers. Example: 0 (number), "0" (string), bMAo= (base64 of [48])
* bytes: should we use hex or base64 or base58 with "b" prefix, eg : bYWJjCg==  is a base64 of  [61, 62, 63]

Reply:
Added.
- For strings, I don't think we should do `"`, just plain. Addresses are strings, it's weird to wrap them in "".
- For bytes, I think hexadecimal makese the most sense.

Reply:
I think base64 is the the most often used format for bytes (in general, eg database, REST etc...)

Reply:
Yeah i think base64 is more common on bytes encoding. Can we go with that?